python/CMakeLists.txt

```diff
@@ -112,35 +112,46 @@ endmacro(py_compile)
 py_compile(${CMAKE_CURRENT_BINARY_DIR} ${pysrc-generated} CPROTON_ARTIFACTS)
 py_compile(${CMAKE_CURRENT_SOURCE_DIR} "${pysrc}" PROTON_ARTIFACTS)
 
-find_program(EPYDOC_EXE epydoc)
-mark_as_advanced (EPYDOC_EXE)
-if (EPYDOC_EXE)
-   foreach (py_src_doc ${pysrc})
-     list(APPEND PY_DOC_FILES "${CMAKE_CURRENT_SOURCE_DIR}/${py_src_doc}")
-   endforeach(py_src_doc)
+
+find_program(SPHINX_EXE_PY2 sphinx-build-2 HINTS "/usr/bin")
+mark_as_advanced (SPHINX_EXE_PY2)
+find_program(SPHINX_EXE_PY3 sphinx-build-3 HINTS "/usr/bin")
+mark_as_advanced (SPHINX_EXE_PY3)
+
+# Version of Python used to build API must match version of Sphinx used to build the docs
+get_filename_component(PYTHON_EXECUTABLE_NAME ${PYTHON_EXECUTABLE} NAME)
+# Try matching for Python 3 first
+if ((SPHINX_EXE_PY3) AND (PYTHON_EXECUTABLE_NAME STREQUAL "python3"))
+    set(BUILD_SPHINX_DOCS TRUE)
```

Comment:
In the event the appropriate Sphinx for the version of Python in use is not installed, this approach will result in an error when using `make docs`:
```
/usr/bin/python: No module named sphinx
make[3]: *** [python/CMakeFiles/docs-py.dir/build.make:57: python/CMakeFiles/docs-py] Error 1
make[2]: *** [CMakeFiles/Makefile2:3866: python/CMakeFiles/docs-py.dir/all] Error 2
make[1]: *** [CMakeFiles/Makefile2:115: CMakeFiles/docs.dir/rule] Error 2
make: *** [Makefile:188: docs] Error 2
```
But it certainly would address the other issues raised above. I'll look for a way to handle the error more gracefully.

Reply:
You missed the module detection section of my suggestion (perhaps because it is so much smaller than what was there before). This sets SPHINX_MODULE_FOUND and you can use this to decide whether or not to try and build the docs.

Reply:
I should have been more precise describing how this error occurred. I did include this section. The `python3-sphinx` module was installed (and found), but not the Python 2 version. The `SPHINX_MODULE_FOUND` variable returns `true`. However, when building qpid-proton with the (default) Python2, this error occurs. If we could find a way to distinguish between Python2 and Python3 versions of Sphinx, then logic can be used to prevent and error.

Reply:
This doesn't make a lot of sense. check_python_module() uses PYTHON_EXECUTABLE to find the module, so if the build is using python2 then the module detection should be using python2 as well. Are you sure you are correctly testing SPHINX_MODULE_FOUND? Or perhaps you're somehow using python directly and through PYTHON_EXECUTABLE?

Reply:
```
check_python_module("sphinx" SPHINX_MODULE_FOUND)
if (SPHINX_MODULE_FOUND)
    add_custom_target(docs-py
        COMMAND ${PN_ENV_SCRIPT} --
        PYTHONPATH=${CMAKE_CURRENT_BINARY_DIR}:${CMAKE_CURRENT_SOURCE_DIR}
        LD_LIBRARY_PATH="${CMAKE_CURRENT_BINARY_DIR}/c"
        ${PYTHON_EXECUTABLE} -m sphinx "${CMAKE_CURRENT_SOURCE_DIR}/docs" "${CMAKE_CURRENT_BINARY_DIR}/docs")
    add_dependencies(docs docs-py)
    install(DIRECTORY "${CMAKE_CURRENT_BINARY_DIR}/docs/"
            DESTINATION "${PROTON_SHARE}/docs/api-py"
            COMPONENT documentation
            OPTIONAL)
endif ()
```
Let me run the test again and report on exactly which versions of Sphinx are installed.

Reply:
```
$ sudo yum list python?-sphinx*
Installed Packages
python3-sphinx.noarch                                                     1:1.8.4-1.fc30                                 @fedora
python3-sphinx-theme-alabaster.noarch                                     0.7.12-3.fc30                                  @fedora
python3-sphinx_rtd_theme.noarch                                           0.4.3-1.fc30                                   @fedora
python3-sphinxcontrib-apidoc.noarch                                       0.2.1-10.fc30                                  @fedora
python3-sphinxcontrib-websupport.noarch                                   1.0.1-12.20180316git.fc30                      @fedora
```
I repeated the test by:
1. Erasing the python2-sphinx package (`dnf erase python2-sphinx`)
2. Running `cmake ..` from the build directory
3. Running `make clean all docs`
but it produced the same results as above. I'll start some debugging...

Reply:
In your invocation of sphinx you do not need (probably should not have) the PYTHONPATH and LB_LIBRARY_PATH environment vars set. I'm not sure this makes a difference, but it certainly clutters the CMake code for no reason.

Reply:
Can you paste the output from cmake as well?

Reply:
By adding debug message() statements prior to the if(), I see:
```
PYTHON_EXECUTABLE=/usr/bin/python
SPHINX_MODULE_FOUND=1
```
with the above packages installed.

Reply:
It looks like a problem with check_python_module(), it returns `SPHINX_MODULE_FOUND=1` with both python2 and python3 versions of sphinx erased.

Reply:
I think I see the problem: You aren't clearing the CMake results between tests. check_python_module() like the standard cmake checks caches the results if it finds the module.
This means that if you rerun the test after erasing something the test is not rerun to find a new result.
This is a completely general cmake behaviour - if you remove python etc. cmake won't discover this and will try to use the previous result.
So you need to test the absence of sphinx with a clean cmake configure run.

Reply:
Agreed. I discovered that initially, I could check for any arbitrary module name, and it would result in a positive check. However, only after blowing away the build directory in its entirety, did correct behavior return, and the detection of sphinx through either Python 2 or 3 worked correctly.

This clears the way to checking in this change. I must look into the environment variable issue. Without both of these, the sphinx build fails, but perhaps they should be in the conf file somewhere.

Reply:
If sphinx has to import the proton module in python to work then I suppose it needs the correct environment for that.
In other words 'import proton' or equivalent would require the environment to be set up correctly to run proton and these environment vars would allow that to work.
A little ugly IMO.

Reply:
http://www.sphinx-doc.org/en/master/usage/extensions/autodoc.html
Yes, it is the autodoc module that imports the target sources and extracts the docstrings automatically. I'll leave the PYTHONPATH and LD_LIBRARY_PATH environment variables in for now. If I can find a way to put them into the config file, I'll do that later, but the config file is not aware of cmake variables, so it will probably involve a cmake write to the config file? That would be ugly, we will see.

Reply:
I think resolving this neatly will involve restructuring the build a little, so this is fine for the moment.